[git] Group the context menu items per Git command #8400

dhuebner opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

Groups menu items in git toolbar menu into sub-menus



#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
Open Source Control view, open toolbar menu. Grouped menu should be shown:
![Bildschirmfoto 2021-04-09 um 15 20 57](https://user-images.githubusercontent.com/454322/114187699-93206180-9948-11eb-9fe4-47b399083989.png)
  

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

kittaakos commented:
Super. Thank you for the patch. Could you please add groups for Pull/Push and keep the order from VS Code?

VS Code:
![Screen Shot 2021-04-12 at 09 15 50](https://user-images.githubusercontent.com/1405703/114358571-56847e00-9b73-11eb-9e03-03e2d6e4ec57.jpg)


Theia:
![Screen Shot 2021-04-12 at 09 15 59](https://user-images.githubusercontent.com/1405703/114358549-51273380-9b73-11eb-9609-22d160abdad9.jpg)

dhuebner commented:
@kittaakos 
I'm not sure how to define groups in submenus, I need to try it out

vince-fugnitto reviewed (commented):
I have noticed the following logs with the changes:

![image](https://user-images.githubusercontent.com/40359487/114398424-c859c900-9b6d-11eb-9a14-eae17972cad2.png)

dhuebner commented on packages/git/src/browser/git-contribution.ts:
```diff
@@ -204,7 +204,32 @@ export namespace GIT_COMMANDS {
     };
 }
+export namespace GIT_MENUS {
+    // Main menu groups
```
Yes, you are right. I will remove it

dhuebner commented:
> I have noticed the following logs with the changes:
> 
> ![image](https://user-images.githubusercontent.com/40359487/114398424-c859c900-9b6d-11eb-9a14-eae17972cad2.png)

@vince-fugnitto 
I didn't noticed that :( 

This is probably because we have PULL and PUSH in favorites group and as submenu entry. Would it be fine to just create the entries with different command id's or is there a way to reference the same command in different menu items?

vince-fugnitto commented:
@dhuebner 

The error happens when we attempt to register the same command `id` multiple times:

https://github.com/eclipse-theia/theia/blob/b722e4de165d0a364b27815c9a68bfdf2d031c4a/packages/core/src/common/command.ts#L189-L198

This happens when we call `registerItem`:

https://github.com/eclipse-theia/theia/blob/b722e4de165d0a364b27815c9a68bfdf2d031c4a/packages/git/src/browser/git-contribution.ts#L541-L545

One solution might need to register a new command which wraps the existing, and is used in the submenu, but there might be others:

```ts
export const PULL_DEFAULT_SUBMENU = {
  id: 'git.pull.default.submenu',
};
```

```ts
registry.registerCommand(GIT_COMMANDS.PULL_DEFAULT_SUBMENU, {
  execute: () => registry.executeCommand(GIT_COMMANDS.PULL_DEFAULT.id),
  isEnabled: () => !!this.repositoryTracker.selectedRepository
});
```

dhuebner commented:
@kittaakos 
Groups are added to the submenu:

![Bildschirmfoto 2021-04-16 um 15 44 41](https://user-images.githubusercontent.com/454322/115033421-b90ab080-9eca-11eb-940e-1944bfca51e0.png)

Unfortunately priorities are lost due to missing `order` property during `renderMoreContextMenu` in tab-bar-toolbar.tsx, I added TODO comments. 

@vince-fugnitto 
I also registered the favorites menu entries (Pull/Push) with new ids like you suggested.

kittaakos commented:
Theia:
![Screen Shot 2021-04-23 at 08 06 42](https://user-images.githubusercontent.com/1405703/115826399-26f33280-a40b-11eb-814d-cb8922eb8485.jpg)

VS Code:
![Screen Shot 2021-04-23 at 08 07 10](https://user-images.githubusercontent.com/1405703/115826419-2e1a4080-a40b-11eb-85c8-f78bd532f7db.jpg)

Besides some ordering differences, the _Changes_ menu groups are in sync 👍. Why did you re-declare the _Stage All Changes_ in the _Pull, Push_ submenu? Is it a bug or feature?

Theia:
![Screen Shot 2021-04-23 at 08 06 47](https://user-images.githubusercontent.com/1405703/115826634-80f3f800-a40b-11eb-93aa-86192374b750.jpg)

VS Code:
![Screen Shot 2021-04-23 at 08 11 54](https://user-images.githubusercontent.com/1405703/115826689-9537f500-a40b-11eb-83e6-0b4b62978ee7.jpg)

dhuebner commented:
@kittaakos 
Thanks for your feedback. I fixed the Fetch command label (Stage All Changes -> Fetch...)  and also moved Merge... from favorites to command group.
![Bildschirmfoto 2021-04-23 um 08 53 21](https://user-images.githubusercontent.com/454322/115831062-63299180-a411-11eb-8723-db20cc119717.png)


Regarding the order, I see any chance to do it properly because of alphabetical sorting due to missing SubMenuOptions here: https://github.com/eclipse-theia/theia/pull/9324/files#diff-07529f987e23173e43dafbf57a0b4b625978849167b48205fc30d3f806feec01

Missing `order` leads to the alphabetical sorting here: https://github.com/eclipse-theia/theia/blob/master/packages/core/src/common/menu.ts#L341

...because the sorting string is the MenuItem label

kittaakos reviewed (approved):
I have verified it on Gitpod, and it worked:
 - there were no warnings in the console due to the duplicate command registration,
 - all menu items were there, and
 - the Git commands I have tried ran successfully.
 
Thank you, @dhuebner 👍 

@vince-fugnitto, could you please check if you're happy with the merge?

vince-fugnitto commented on examples/README.md:
```diff
@@ -0,0 +1,5 @@
+-- Debug browser app
+
+1. `yarn --cwd ./packages/git watch` (to watch TS changes in src -> compiles to JS into the lib folder)
+2. `yarn --cwd ./examples/browser watch` (to watch for webpack. whenever the previous step modifies the packages in the lib folder, we regenerate the final bundle.js from the .js files for the app)
+3. `yarn --cwd ./examples/browser start` to start the app.
```
@dhuebner I believe this file can be reverted?

dhuebner replied:
@vince-fugnitto
Sure. I just thought it might be useful for other contributors.

vince-fugnitto replied:
@dhuebner thank you for the initiative, however, documentation regarding development is already covered in [developping.md](https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md) which is also linked in the readme. If we are to improve the documentation regarding development, starting or debugging the examples it should be added to [developing.md](https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md) as it would be easier to maintain and discover.

vince-fugnitto reviewed (commented):
I confirmed that the menu is correctly populated and organized when testing :+1:

kittaakos commented:
@dhuebner, can you please squash your commits? Thank you!

dhuebner commented:
@kittaakos 
Done!
